segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/exts3/ExtS3Storage.java

```diff
@@ -0,0 +1,420 @@
+        return retHandle;
+    }
+
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        StreamSegmentInformation info = syncGetStreamSegmentInfo(streamSegmentName);
+        ExtS3SegmentHandle retHandle;
+        if (info.isSealed()) {
+            retHandle = ExtS3SegmentHandle.getReadHandle(streamSegmentName);
+        } else {
+            retHandle = ExtS3SegmentHandle.getWriteHandle(streamSegmentName);
+        }
+
+        LoggerHelpers.traceLeave(log, "openWrite", traceId);
+        return retHandle;
+    }
+
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) throws IOException, StreamSegmentNotExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle.getSegmentName(), offset, bufferOffset, length);
+
+        if (offset < 0 || bufferOffset < 0 || length < 0) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+
+        try (InputStream reader = client.readObjectStream(config.getExts3Bucket(),
+                config.getExts3Root() + handle.getSegmentName(), Range.fromOffsetLength(offset, length))) {
+
+            if (reader == null) {
+                throw new StreamSegmentNotExistsException(handle.getSegmentName());
+            }
+
+            int originalLength = length;
+
+            int bytesRead = StreamHelpers.readAll(reader, buffer, bufferOffset, length);
+
+            LoggerHelpers.traceLeave(log, "read", traceId, bytesRead);
+            return originalLength;
+        }
+    }
+
+    private StreamSegmentInformation syncGetStreamSegmentInfo(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+        S3ObjectMetadata result = client.getObjectMetadata(config.getExts3Bucket(),
+                config.getExts3Root() + streamSegmentName);
+
+        AccessControlList acls = client.getObjectAcl(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName);
+
+        boolean canWrite = false;
+        canWrite = acls.getGrants().stream().filter((grant) -> {
+            return grant.getPermission().compareTo(Permission.WRITE) >= 0;
+        }).count() > 0;
+
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName,
+                result.getContentLength(), !canWrite, false,
+                new ImmutableDate(result.getLastModified().toInstant().toEpochMilli()));
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        GetObjectResult<InputStream> result = null;
+        try {
+            result = client.getObject(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName);
```

Comment:
Yes, but the implementation almost certainly assumes that some will be read, so while it may not have any in memory yet to return it has almost certainly issued a request to the server to read by the time the result is returned.